File: modules/ia_cardlink/ios/ia_cardlink/Sources/ia_cardlink/Events/CardLinkEventSender.swift
import Flutter
import IACore
import IACardLink

@MainActor
class CardLinkEventSender {
    private var methodChannel: FlutterMethodChannel?

    func setMethodChannel(_ channel: FlutterMethodChannel?) {
        self.methodChannel = channel
    }

    func sendConsentEvent(accepted: Bool, phoneNumber: String?) {
        let status = accepted ? "accepted" : "declined"
        methodChannel?.invokeMethod(
            "cardLinkConsentEvent",
            arguments: ["status": status]
        )
    }

    func sendSessionCreatedEvent(session: CardLinkOutputAction.CardLinkSession) {
        let sessionDict: [String: Any] = [
            "cardSessionId": session.cardSessionId,
            "sessionExpireTimestamp": session.sessionExpiresAt
        ]
        methodChannel?.invokeMethod(
            "cardLinkSessionCreated",
            arguments: ["session": sessionDict]
        )
    }

    func sendPrescriptionsRedeemedEvent(prescriptions: String) {
        methodChannel?.invokeMethod(
            "cardLinkPrescriptionsRedeemed",
            arguments: ["prescriptions": prescriptions]
        )
    }

    func sendEvent(_ event: CardLinkEventType) {
        methodChannel?.invokeMethod(
            "cardLinkEvent",
            arguments: ["event": event.rawValue]
        )
    }

    func sendCardSavedEvent(cards: [CardData]) {
        let formattedCards = cards.map { card in
            return [
                "id": card.id,
                "name": card.name,
                "canNumber": card.canNumber,
                "phoneNumber": card.phoneNumber
            ]
        }
        methodChannel?.invokeMethod(
            "cardLinkEvent",
            arguments: ["event": "cardSaved", "cards": formattedCards]
        )
    }

    func sendFailedToInitializeEvent(error: CardLinkOutputAction.InitializationError) {
        let reason: String
        switch error {
        case .missingAuthenticationKey:
            reason = "Missing authentication key."
        case .missingPharmacyID:
            reason = "Missing pharmacy ID."
        case .failedToAuthenticate:
            reason = "Failed to authenticate, please check your authentication key and bundle id."
        default:
            reason = "Unknown error"
        }
        methodChannel?.invokeMethod(
            "cardLinkEvent",
            arguments: [
                "event": "failedToInitialize",
                "errorMessage": "CardLink SDK failed to initialize: \(reason)"
            ]
        )
    }

    func sendAnalyticsEvent(event: String) {
        methodChannel?.invokeMethod(
            "cardLinkAnalyticsEvent",
            arguments: ["analyticEvent": event]
        )
    }
}

File: modules/ia_cardlink/android/src/main/kotlin/de/ihreapotheken/appsdk/flutter/card_link/events/CardLinkEventSender.kt
package de.ihreapotheken.appsdk.flutter.card_link.events

import android.app.Activity
import android.os.Handler
import android.os.Looper
import de.ihreapotheken.appsdk.flutter.card_link.models.toMap
import de.ihreapotheken.sdk.cardlink.domain.model.CardLinkSession
import de.ihreapotheken.sdk.cardlink.domain.model.insurancecard.InsuranceCard
import io.flutter.plugin.common.MethodChannel

class CardLinkEventSender(
    private val activityProvider: () -> Activity?,
    private val methodChannelProvider: () -> MethodChannel?
) {
    private val mainHandler = Handler(Looper.getMainLooper())

    fun sendConsentEvent(accepted: Boolean) {
        val status = if (accepted) "accepted" else "declined"
        mainHandler.post {
            methodChannelProvider()?.invokeMethod(
                "cardLinkConsentEvent",
                mapOf("status" to status)
            )
        }
    }

    fun sendSessionCreatedEvent(session: CardLinkSession) {
        mainHandler.post {
            methodChannelProvider()?.invokeMethod(
                "cardLinkSessionCreated",
                mapOf("session" to session.toMap())
            )
        }
    }

    fun sendPrescriptionsRedeemedEvent(prescriptions: String) {
        mainHandler.post {
            methodChannelProvider()?.invokeMethod(
                "cardLinkPrescriptionsRedeemed",
                mapOf("prescriptions" to prescriptions)
            )
        }
    }

    fun sendGoToCartEvent() {
        mainHandler.post {
            methodChannelProvider()?.invokeMethod(
                "cardLinkEvent",
                mapOf("event" to "goToCart")
            )
        }
    }

    fun sendOpenTermsAndConditionsEvent() {
        mainHandler.post {
            methodChannelProvider()?.invokeMethod(
                "cardLinkEvent",
                mapOf("event" to "openTermsAndConditions")
            )
        }
    }

    fun sendCardSavedEvent(card: InsuranceCard) {
        val formattedCard = listOf(
            mapOf(
                "id" to card.userId,
                "name" to card.userId,
                "canNumber" to card.can,
                "phoneNumber" to card.phoneNumber
            )
        )

        mainHandler.post {
            methodChannelProvider()?.invokeMethod(
                "cardLinkEvent",
                mapOf("event" to "cardSaved", "cards" to formattedCard)
            )
        }
    }

    fun sendAnalyticsEvent(analyticEvent: String) {
        mainHandler.post {
            methodChannelProvider()?.invokeMethod(
                "cardLinkAnalyticsEvent",
                mapOf("analyticEvent" to analyticEvent)
            )
        }
    }

    fun sendFailedToInitializeEvent() {
        mainHandler.post {
            methodChannelProvider()?.invokeMethod(
                "cardLinkEvent",
                mapOf("event" to "failedToInitialize")
            )
        }
    }

    fun sendCardLinkResultEvent(result: String?) {
        mainHandler.post {
            methodChannelProvider()?.invokeMethod(
                "cardLinkEvent",
                mapOf("event" to "cardLinkResult", "result" to result)
            )
        }
    }

    fun sendCardLinkCanceledEvent() {
        mainHandler.post {
            methodChannelProvider()?.invokeMethod(
                "cardLinkEvent",
                mapOf("event" to "cardLinkResult", "result" to "canceled")
            )
        }
    }
}
